### PR TITLE
fix: Solving `make run` errors under Go vendor mode.

### PR DESCRIPTION
**Pull request by village-way**
#### What type of PR is this?

kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
@alculquicondor

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -327,7 +327,7 @@ golangci-lint: ## Download golangci-lint locally if necessary.
 CONTROLLER_GEN = $(PROJECT_DIR)/bin/controller-gen
 .PHONY: controller-gen
 controller-gen: ## Download controller-gen locally if necessary.
-	@GOBIN=$(PROJECT_DIR)/bin GO111MODULE=on $(GO_CMD) install sigs.k8s.io/controller-tools/cmd/controller-gen
+	@GOBIN=$(PROJECT_DIR)/bin GO111MODULE=on $(GO_CMD) install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.14.0
```
Comment (trasc) [inline Makefile:330]: `controller-gen` should use the version set in go.mod

**Review by trasc — COMMENTED** [review]
Do  you have a specific use-case for this? 

Depending on what you want to do, in my opinion , it's easier to just  `install` and `deploy`

File: Makefile (hunk 1 above)
Comment (alculquicondor) [reply Makefile:330]: Indeed. Please revert this line, as the version is locked in go.mod  https://github.com/kubernetes-sigs/kueue/pull/1720

File: hack/update-codegen.sh [hunk 2]
```diff
@@ -19,11 +19,14 @@ set -o nounset
 set -o pipefail
 
 GO_CMD=${1:-go}
-KUEUE_ROOT=$(realpath $(dirname ${BASH_SOURCE[0]})/..)
-CODEGEN_PKG=$($GO_CMD list -m -f "{{.Dir}}" k8s.io/code-generator)
-
-cd $(dirname ${BASH_SOURCE[0]})/..
-
+KUEUE_ROOT=$(realpath $(dirname "${BASH_SOURCE[0]})"/..))
+if [ -d "vendor" ]; then
```
Comment (alculquicondor) [inline hack/update-codegen.sh:23]: why do you have a vendor folder?

File: Makefile (hunk 1 above)
Comment (village-way) [reply Makefile:330]: > Indeed. Please revert this line, as the version is locked in go.mod #1720
I am a novice in Go and Kubernetes development. I wanted to address the following issue, which is why I changed that line of code.
image

**Comment by village-way** [comment]
> Do you have a specific use-case for this?
> 
> Depending on what you want to do, in my opinion , it's easier to just `install` and `deploy`

I just want to start the kueue main.go process and debug the code in the goland, so want to run that in `make run` with flags or in the goland Run/Debug configurations

File: hack/update-codegen.sh (hunk 2 above)
Comment (village-way) [reply hack/update-codegen.sh:23]: > why do you have a vendor folder?


"The 'go mod vendor' command automatically generates the 'vendor' directory. If this command is not used to copy dependencies to the project space, the GoLang editor cannot automatically locate dependencies such as 'sigs.k8s.io/controller-runtime', resulting in red error indicators indicating that the dependencies cannot be found."

**Comment by tenzen-y** [comment]
> > Do you have a specific use-case for this?
> > Depending on what you want to do, in my opinion , it's easier to just `install` and `deploy`
> 
> I just want to start the kueue main.go process and debug the code in the goland, so want to run that in `make run` with flags or in the goland Run/Debug configurations

I'm not sure why we should have the vendor if we use Goland.

**Comment by alculquicondor** [comment]
Can you make sure you are using go v1.21 or newer and that you enable go modules support in golang? https://www.jetbrains.com/help/go/create-a-project-with-go-modules-integration.html#using-go-modules-in-your-project

You should not run `go mod vendor`

**Comment by village-way** [comment]
> Can you make sure you are using go v1.21 or newer and that you enable go modules support in golang? https://www.jetbrains.com/help/go/create-a-project-with-go-modules-integration.html#using-go-modules-in-your-project
> 
> You should not run `go mod vendor`
this is the go env
```
GO111MODULE='on'
GOARCH='amd64'
GOBIN='/Users/devine/go'
GOCACHE='/Users/devine/Library/Caches/go-build'
GOENV='/Users/devine/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/Users/devine/go/pkg/mod'
GONOPROXY='gitlab.*.com'
GONOSUMDB='gitlab.*.com'
GOOS='darwin'
GOPATH='/Users/devine/go'
GOPRIVATE='gitlab.*.com'
GOPROXY='https://goproxy.cn,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/usr/local/go/pkg/tool/darwin_amd64'
GOVCS=''
GOVERSION='go1.22.1'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='clang'
CXX='clang++'
CGO_ENABLED='1'
GOMOD='/Users/devine/GolandProjects/kueue/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/fm/sp8j2q9d5q55x1rdzlywfzs00000gp/T/go-build1822828399=/tmp/go-build -gno-record-gcc-switches -fno-common'

```
if not run `go mod vendor` everythings is ok, after run `go mod vendor` the make run this command, will occurs `cannot find module providing package sigs.k8s.io/controller-tools/cmd/controller-gen: import lookup disabled by -mod=vendor
        (Go version in go.mod is at least 1.14 and vendor directory exists.)
make: *** [controller-gen] Error 1
` that's why i add the version for the controller-gen in the script dependencies. Only run `go mod vendor`, the goland can find the dependencies. The config of goland is 
image

**Comment by village-way** [comment]
> > > Do you have a specific use-case for this?
> > > Depending on what you want to do, in my opinion , it's easier to just `install` and `deploy`
> > 
> > 
> > I just want to start the kueue main.go process and debug the code in the goland, so want to run that in `make run` with flags or in the goland Run/Debug configurations
> 
> I'm not sure why we should have the vendor if we use Goland.


Every time GoLand is opened, it executes go list. However, this go list command fails, and it's unclear if this is the cause. Executing go mod vendor is just to allow GoLand to find dependencies without prompting that they can't be found.

image

**Comment by tenzen-y** [comment]
> > > > Do you have a specific use-case for this?
> > > > Depending on what you want to do, in my opinion , it's easier to just `install` and `deploy`
> > > 
> > > 
> > > I just want to start the kueue main.go process and debug the code in the goland, so want to run that in `make run` with flags or in the goland Run/Debug configurations
> > 
> > 
> > I'm not sure why we should have the vendor if we use Goland.
> 
> Every time GoLand is opened, it executes go list. However, this go list command fails, and it's unclear if this is the cause. Executing go mod vendor is just to allow GoLand to find dependencies without prompting that they can't be found.
> 
> image

That problem could be fixed by https://github.com/kubernetes-sigs/kueue/issues/1345.

**Comment by village-way** [comment]
> > > > > Do you have a specific use-case for this?
> > > > > Depending on what you want to do, in my opinion , it's easier to just `install` and `deploy`
> > > > 
> > > > 
> > > > I just want to start the kueue main.go process and debug the code in the goland, so want to run that in `make run` with flags or in the goland Run/Debug configurations
> > > 
> > > 
> > > I'm not sure why we should have the vendor if we use Goland.
> > 
> > 
> > Every time GoLand is opened, it executes go list. However, this go list command fails, and it's unclear if this is the cause. Executing go mod vendor is just to allow GoLand to find dependencies without prompting that they can't be found.
> > image
> 
> That problem could be fixed by #1345.

Thank you very much,  are according to #1345 , this problem can be fixed and we don't need run `go mod vendor`, so why not add this lines to go.mod file ? @tenzen-y 
```
replace (
	k8s.io/dynamic-resource-allocation => k8s.io/dynamic-resource-allocation v0.28.3
	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.28.3
	k8s.io/kubectl => k8s.io/kubectl v0.28.3
	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.28.3
)
```

**Comment by tenzen-y** [comment]
> > > > > > Do you have a specific use-case for this?
> > > > > > Depending on what you want to do, in my opinion , it's easier to just `install` and `deploy`
> > > > > 
> > > > > 
> > > > > I just want to start the kueue main.go process and debug the code in the goland, so want to run that in `make run` with flags or in the goland Run/Debug configurations
> > > > 
> > > > 
> > > > I'm not sure why we should have the vendor if we use Goland.
> > > 
> > > 
> > > Every time GoLand is opened, it executes go list. However, this go list command fails, and it's unclear if this is the cause. Executing go mod vendor is just to allow GoLand to find dependencies without prompting that they can't be found.
> > > image
> > 
> > 
> > That problem could be fixed by #1345.
> 
> Thank you very much, according to #1345 , this problem can be fixed and we don't need run `go mod vendor`, so why not add this lines to go.mod file ? @tenzen-y
> 
> ```
> replace (
> 	k8s.io/dynamic-resource-allocation => k8s.io/dynamic-resource-allocation v0.28.3
> 	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.28.3
> 	k8s.io/kubectl => k8s.io/kubectl v0.28.3
> 	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.28.3
> )
> ```

I don't think so since after #1345 is fixed, we don't need such imports.
We're working on fixing it in the autoscaler repository.

**Comment by alculquicondor** [comment]
@tenzen-y do you think we are close to merging the split of cluster-autoscaler packages?
If it's taking too long, maybe it's worth introducing the `replace` for now?

**Comment by tenzen-y** [comment]
> do you think we are close to merging the split of cluster-autoscaler packages?
> If it's taking too long, maybe it's worth introducing the `replace` for now?

All tasks have already been done. I'm waiting for approval from @x13n and @mwielgus. Both approvals are necessary.

**Comment by tenzen-y** [comment]
> > do you think we are close to merging the split of cluster-autoscaler packages?
> > If it's taking too long, maybe it's worth introducing the `replace` for now?
> 
> All tasks have already been done. I'm waiting for approval from @x13n (for CA approval) and @mwielgus (for autoscaler repository approval). Both approvals are necessary.

Or only approval from @mwielgus should be fine since @x13n already approved that PR.

**Comment by tenzen-y** [comment]
> > > > > > > > Do you have a specific use-case for this?
> > > > > > > > Depending on what you want to do, in my opinion , it's easier to just `install` and `deploy`
> > > > > > > 
> > > > > > > 
> > > > > > > I just want to start the kueue main.go process and debug the code in the goland, so want to run that in `make run` with flags or in the goland Run/Debug configurations
> > > > > > 
> > > > > > 
> > > > > > I'm not sure why we should have the vendor if we use Goland.
> > > > > 
> > > > > 
> > > > > Every time GoLand is opened, it executes go list. However, this go list command fails, and it's unclear if this is the cause. Executing go mod vendor is just to allow GoLand to find dependencies without prompting that they can't be found.
> > > > > image
> > > > 
> > > > 
> > > > That problem could be fixed by #1345.
> > > 
> > > 
> > > Thank you very much, according to #1345 , this problem can be fixed and we don't need run `go mod vendor`, so why not add this lines to go.mod file ? @tenzen-y
> > > ```
> > > replace (
> > > 	k8s.io/dynamic-resource-allocation => k8s.io/dynamic-resource-allocation v0.28.3
> > > 	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.28.3
> > > 	k8s.io/kubectl => k8s.io/kubectl v0.28.3
> > > 	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.28.3
> > > )
> > > ```
> > 
> > 
> > I don't think so since after #1345 is fixed, we don't need such imports. We're working on fixing it in the autoscaler repository.
> 
> I saw the PR you mentioned at [kubernetes/autoscaler#6315](https://github.com/kubernetes/autoscaler/pull/6315). I'm not very familiar with the Go module dependencies, but I roughly understand that the issue arises because the clusterAutoscale is trying to reuse the code from Schedule without it being a separate Go module. However, I'm still quite confused about this area. Could you please help explain why this problem occurred? I'm looking forward to your reply. ♥

k/k repository is not assumed to be used as a go module. This issue might be helpful for you to understand reasons: https://github.com/kubernetes/kubernetes/issues/79384

**Comment by village-way** [comment]
> > > > > > > > > Do you have a specific use-case for this?
> > > > > > > > > Depending on what you want to do, in my opinion , it's easier to just `install` and `deploy`
> > > > > > > > 
> > > > > > > > 
> > > > > > > > I just want to start the kueue main.go process and debug the code in the goland, so want to run that in `make run` with flags or in the goland Run/Debug configurations
> > > > > > > 
> > > > > > > 
> > > > > > > I'm not sure why we should have the vendor if we use Goland.
> > > > > > 
> > > > > > 
> > > > > > Every time GoLand is opened, it executes go list. However, this go list command fails, and it's unclear if this is the cause. Executing go mod vendor is just to allow GoLand to find dependencies without prompting that they can't be found.
> > > > > > image
> > > > > 
> > > > > 
> > > > > That problem could be fixed by #1345.
> > > > 
> > > > 
> > > > Thank you very much, according to #1345 , this problem can be fixed and we don't need run `go mod vendor`, so why not add this lines to go.mod file ? @tenzen-y
> > > > ```
> > > > replace (
> > > > 	k8s.io/dynamic-resource-allocation => k8s.io/dynamic-resource-allocation v0.28.3
> > > > 	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.28.3
> > > > 	k8s.io/kubectl => k8s.io/kubectl v0.28.3
> > > > 	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.28.3
> > > > )
> > > > ```
> > > 
> > > 
> > > I don't think so since after #1345 is fixed, we don't need such imports. We're working on fixing it in the autoscaler repository.
> > 
> > 
> > I saw the PR you mentioned at [kubernetes/autoscaler#6315](https://github.com/kubernetes/autoscaler/pull/6315). I'm not very familiar with the Go module dependencies, but I roughly understand that the issue arises because the clusterAutoscale is trying to reuse the code from Schedule without it being a separate Go module. However, I'm still quite confused about this area. Could you please help explain why this problem occurred? I'm looking forward to your reply. ♥
> 
> k/k repository is not assumed to be used as a go module. This issue might be helpful for you to understand reasons: [kubernetes/kubernetes#79384](https://github.com/kubernetes/kubernetes/issues/79384)

thanks, this pr can be closed as the related issue solved.

**Comment by tenzen-y** [comment]
In favor of #1872 
/close

**Pull request closed without merging**